### PR TITLE
(NFC) Enhance \Civi\Token\AbstractTokenSubscriber::prefetch() 

### PR DESCRIPTION
**Pull request by aydun**
Enhance \Civi\Token\AbstractTokenSubscriber::prefetch() to receive list of active tokens

Update CRM_Mailing_Tokens::prefetch() to match

Another extraction from #13174 and #12012

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
I agree this is the only core place where that function is overridden

I guess the only question is whether anything outside core overrides it - since the class is in the Civi dir it could be argued it's a formal interface & we support extending it - the alternative would be for $activeTokens to be a class property & it would be set prior to calling prefetch - which would mean no signature change

@totten do you have your universe tool keyed up to check that?

**Comment by seamuslee001** [comment]
I have just checked Flexmailer and it doesn't use the prefetch function does implement other things i thin but not prefetch

File: CRM/Mailing/Tokens.php [hunk 1]
```diff
@@ -72,7 +72,7 @@ public function checkActive(\Civi\Token\TokenProcessor $processor) {
    * @return array
    * @throws \Exception
    */
-  public function prefetch(\Civi\Token\Event\TokenValueEvent $e) {
+  public function prefetch(\Civi\Token\Event\TokenValueEvent $e, $activeTokens) {
```
Comment (totten) [inline CRM/Mailing/Tokens.php]: Docblock? As someone browsing the code, it'd help to have an example of the data that will be passed in here.

**Comment by totten** [comment]
> I guess the only question is whether anything outside core overrides it (*prefetch()*)

Good question.

> @totten do you have your universe tool keyed up to check that?

My copy is little dated but nothing shows up. (*I'm half way through updating and haven't found anything relevant in the updates... unless there's no more feedback from me within the hour, you can asume nothing else came up.*)

> I have just checked Flexmailer and it doesn't use the prefetch function does implement other things i thin but not prefetch

Correct. FlexMailer calls `TokenProcessor`  and asks it to render a message... but otherwise it doesn't care *how* the tokens are populated or rendered, so it doesn't need to know anything about `prefetch()`.

**Comment by aydun** [comment]
Thanks all. On reflection maybe making it a class property is better.  I've made changes and pushed those to #13174 so I'll close this.

**Pull request closed without merging**